Print desul and mdspan versions at configure time and in print_configuration

tretre91 opened the pull request:
This PR addresses the remaining items in https://github.com/kokkos/kokkos/issues/7849, that is printing the commit sha of the internal copies of desul and mdspan at configure time and adding them to the output of `print_configuration`.

The output at configure time looks like this when using the internal desul and mdspan
```
-- Architectures:
-- Using bundled desul_atomics copy (desul/desul@fd8da8299519d7ecde91a1c11a3ac8469ce29cac)
-- Performing Test KOKKOS_LINK_OPTIONS_CHECK
-- Performing Test KOKKOS_LINK_OPTIONS_CHECK - Success
-- Experimental mdspan support is enabled
-- Using bundled mdspan copy (kokkos/mdspan@0e4eba59ad53451784b3839bdf98b25fa2d64931)
```
And for the external ones
```
-- Architectures:
-- Using external desul_atomics install found at:
--   ~/Travail/kokkos/desul/desul/install/gnu_serial/lib/cmake/desul/
-- Performing Test KOKKOS_LINK_OPTIONS_CHECK
-- Performing Test KOKKOS_LINK_OPTIONS_CHECK - Success
-- Experimental mdspan support is enabled
-- Using external mdspan
```
The output for `print_configuration` looks like this:
- when using the internal desul and mdspan:
  ```
    Kokkos Version: 4.7.99
  Compiler:
    KOKKOS_COMPILER_GNU: 1330
  Architecture:
    CPU architecture: none
    Default Device: Serial
    GPU architecture: none
    platform: 64bit
  Atomics:
    desul atomics version: bundled (desul/desul@fd8da8299519d7ecde91a1c11a3ac8469ce29cac)
  View:
    mdspan: enabled
    mdspan version: bundled (kokkos/mdspan@0e4eba59ad53451784b3839bdf98b25fa2d64931)
  Vectorization:
    KOKKOS_ENABLE_PRAGMA_IVDEP: no
    KOKKOS_ENABLE_PRAGMA_LOOPCOUNT: no
    KOKKOS_ENABLE_PRAGMA_UNROLL: no
    KOKKOS_ENABLE_PRAGMA_VECTOR: no
  Memory:
  Options:
    KOKKOS_ENABLE_ASM: yes
    KOKKOS_ENABLE_CXX17: no
    KOKKOS_ENABLE_CXX20: yes
    KOKKOS_ENABLE_CXX23: no
    KOKKOS_ENABLE_CXX26: no
    KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK: no
    KOKKOS_ENABLE_HWLOC: no
    KOKKOS_ENABLE_LIBDL: yes
  Host Serial Execution Space:
    KOKKOS_ENABLE_SERIAL: yes
  
  Serial Runtime Configuration:
  ```
- when using external desul and mdspan:
  ```
  Atomics:
    desul atomics version: external
  View:
    mdspan: enabled
    mdspan version: external
  ```
- when using the legacy view:
  ```
  Atomics:
    desul atomics version: bundled (desul/desul@fd8da8299519d7ecde91a1c11a3ac8469ce29cac)
  View:
    mdspan: disabled
  ```

tretre91 committed:
Print desul and mdspan versions at configure time and in print_configuration
- if using the internal desul and/or mdspan, print the corresponding
  commit at configure time and add that information to the output of
  print_configuration()

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

dalg24 commented on core/src/CMakeLists.txt:
```diff
@@ -127,8 +127,12 @@ if(NOT desul_FOUND)
   )
 
-  message(STATUS "Using internal desul_atomics copy")
+  set(KOKKOS_INTERNAL_DESUL_VERSION "desul/desul@fd8da8299519d7ecde91a1c11a3ac8469ce29cac")
```
How do we make sure we keep these hash in sync?

tretre91 replied:
Ideally we would have these hashes defined in a single location, and have the version checks and the cmake script both read from this location, it is possible on the cmake side, I have to check how to do this for github actions.

In case it's not possible, a less robust way would be to add a check in {desul,mdspan}-version-check.yml to grep core/src/CMakeLists.txt and check that the correct hashes are contained in the file

dalg24 commented on core/src/CMakeLists.txt:
```diff
@@ -164,7 +169,13 @@ if(Kokkos_ENABLE_IMPL_MDSPAN)
+    global_set(KOKKOS_MDSPAN_VERSION "internal (${KOKKOS_INTERNAL_MDSPAN_VERSION})")
+    message(
+      STATUS
+        "Using internal mdspan directory ${KOKKOS_SOURCE_DIR}/tpls/mdspan/include (${KOKKOS_INTERNAL_MDSPAN_VERSION})"
```
```suggestion
        "Using internal mdspan copy (${KOKKOS_INTERNAL_MDSPAN_VERSION})"
```

dalg24 reviewed (commented):
Thanks for working on this.

Did you put thoughts on how we would ensure that we do not forget to update these hash when we update https://github.com/kokkos/kokkos/blob/7a516890899a0b0f80393915e27b5880eebf745e/.github/workflows/desul-version-check.yml#L19 and https://github.com/kokkos/kokkos/blob/7a516890899a0b0f80393915e27b5880eebf745e/.github/workflows/mdspan-version-check.yml#L19 ?

tretre91 committed:
Update core/src/CMakeLists.txt

Co-authored-by: Damien L-G <dalg24+github@gmail.com>
Signed-off-by: Trévis Morvany <63788850+tretre91@users.noreply.github.com>

dalg24 commented on core/src/CMakeLists.txt:
```diff
@@ -164,7 +169,13 @@ if(Kokkos_ENABLE_IMPL_MDSPAN)
+    message(
+      STATUS
+        "Using internal mdspan copy (${KOKKOS_INTERNAL_MDSPAN_VERSION})"
+    )
```
Sorry the format was wrong
```suggestion
    message(STATUS "Using internal mdspan copy (${KOKKOS_INTERNAL_MDSPAN_VERSION})")
```

tretre91 committed:
Update core/src/CMakeLists.txt

Co-authored-by: Damien L-G <dalg24+github@gmail.com>
Signed-off-by: Trévis Morvany <63788850+tretre91@users.noreply.github.com>

tretre91 committed:
Merge branch 'develop' into print_mdspan_desul_versions

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Read the hashes from the hash files

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

romintomasetti commented on cmake/KokkosCore_config.h.in:
```diff
@@ -72,6 +72,10 @@
 #cmakedefine KOKKOS_ENABLE_ONEDPL
 #cmakedefine KOKKOS_ENABLE_ROCTHRUST
 
+/* Embedded dependencies */
+#define KOKKOS_IMPL_DESUL_VERSION "@KOKKOS_DESUL_VERSION@"
+#define KOKKOS_IMPL_MDSPAN_VERSION "@KOKKOS_MDSPAN_VERSION@"
+
```
What if a user disables `desul` ? Do we still want the `KOKKOS_IMPL_DESUL_VERSION` to be defined ?

tretre91 replied:
I don't think it is an issue if it is defined, as it is an `IMPL` macro anyway, however I can make changes to only print it if atomics are enabled, the same way I print the mdspan version only if we're not using the legacy view

dalg24 replied:
We cannot disable desul any more.
As for mdspan we could say `"disabled"` for now.  Longer term we will remove that option.

dalg24 commented on core/src/CMakeLists.txt:
```diff
@@ -127,8 +127,12 @@ if(NOT desul_FOUND)
-  message(STATUS "Using internal desul_atomics copy")
+  file(STRINGS "${KOKKOS_SOURCE_DIR}/scripts/desul-hash.txt" KOKKOS_DESUL_HASH)
+  set(KOKKOS_INTERNAL_DESUL_VERSION "desul/desul@${KOKKOS_DESUL_HASH}")
+  global_set(KOKKOS_DESUL_VERSION "internal (${KOKKOS_INTERNAL_DESUL_VERSION})")
```
Should we say "bundled" (instead of internal)?  "embedded"?  "in-tree"?

tretre91 replied:
I would go for "bundled"

tretre91 committed:
Modify the mdspan entry in print_configuration

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Change wording "internal" -> "bundled"

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

dalg24 reviewed (commented):
In the current version we get 
```
-- Using bundled desul_atomics copy (desul/desul@fd8da8299519d7ecde91a1c11a3ac8469ce29cac)
```
at configuration time,
```
#define KOKKOS_IMPL_DESUL_VERSION "bundled (desul/desul@fd8da8299519d7ecde91a1c11a3ac8469ce29cac)"
```
in `KokkosCore_config.h`, and
```
Atomics:
  desul atomics version: bundled (desul/desul@fd8da8299519d7ecde91a1c11a3ac8469ce29cac)
```
when printing the configuration.
With an external desul, it would be
```
-- Using external desul_atomics install found at:
--   /path/to/desul/install/lib/cmake/desul/
```
```
#define KOKKOS_IMPL_DESUL_VERSION ""
```
```
Atomics:
  desul atomics version: external
```
I would prefer if we changed the macro definition to
```
#define KOKKOS_IMPL_DESUL_VERSION "d8da8299519d7ecde91a1c11a3ac8469ce29cac"
```
that is only the git commit hash, and
```
#define KOKKOS_IMPL_DESUL_VERSION "unknown"
```
or similar rather than an empty string if using an external diesel.


As for mdspan, I would define the impl version macro to be `"not applicable"` when not enabled.

For completeness, we currently have
```
-- Using bundled mdspan copy (kokkos/mdspan@0e4eba59ad53451784b3839bdf98b25fa2d64931)
```
```
 #define KOKKOS_IMPL_MDSPAN_VERSION "bundled (kokkos/mdspan@0e4eba59ad53451784b3839bdf98b25fa2d64931)"
```
```
View:
  mdspan: enabled
  mdspan version: bundled (kokkos/mdspan@0e4eba59ad53451784b3839bdf98b25fa2d64931)
```
or
```
-- Experimental mdspan support is enabled
-- Using external mdspan
```
```
 #define KOKKOS_IMPL_MDSPAN_VERSION ""
```
```
View:
  mdspan: enabled
  mdspan version: external
```
or
```
(nothing)
```
```
 #define KOKKOS_IMPL_MDSPAN_VERSION ""
```
```
View:
  mdspan: disabled
```

tretre91 committed:
Always define the VERSION macros
- the KOKKOS_IMPL_{MDSPAN,DESUL}_VERSION macros now contain the commit
  hash if using the bundled copies, otherwise they are defined to "unknown"
- the mdspan macro is defined to "not applicable" and printed if the
  legacy view is enabled

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 commented:
Updated to address the review comment.
The message stays the same at configure time, in `KokkosCore_config.h` we now get
```
#define KOKKOS_IMPL_DESUL_VERSION "fd8da8299519d7ecde91a1c11a3ac8469ce29cac"
#define KOKKOS_IMPL_MDSPAN_VERSION "0e4eba59ad53451784b3839bdf98b25fa2d64931"
```
using the bundled copies,
```
#define KOKKOS_IMPL_DESUL_VERSION "unknown"
#define KOKKOS_IMPL_MDSPAN_VERSION "unknown"
```
using external versions, and
```
#define KOKKOS_IMPL_MDSPAN_VERSION "not applicable"
```
for mdspan when using the legacy view

When printing the configuration, for desul (bundled)
```
Atomics:
  desul atomics version: fd8da8299519d7ecde91a1c11a3ac8469ce29cac
```
or (external)
```
Atomics:
  desul atomics version: unknown
```

and for mdspan (bundled)
```
View:
  mdspan: enabled
  mdspan version: 0e4eba59ad53451784b3839bdf98b25fa2d64931
```
or (external)
```
View:
  mdspan: enabled
  mdspan version: unknown
```
or (legacy view)
```
View:
  mdspan: disabled
  mdspan version: not applicable
```

dalg24 commented:
Plz add an entry to the 5.0 changelog under enhancements.
